File: backend/services/search_service.py
import yfinance as yf
from yfinance import Search
from difflib import SequenceMatcher
from models.stock import SearchResult
from config import MAX_SEARCH_RESULTS

# Exchange suffix mappings for global exchanges
EXCHANGE_SUFFIXES = {
    "NYSE": "",
    "NASDAQ": "",
    "LSE": ".L",
    "Euronext Paris": ".PA",
    "Euronext Amsterdam": ".AS",
    "Deutsche Boerse": ".DE",
    "OMX Copenhagen": ".CO",
    "OMX Stockholm": ".ST",
    "OMX Helsinki": ".HE",
    "Tokyo": ".T",
    "Hong Kong": ".HK",
    "Shanghai": ".SS",
    "Shenzhen": ".SZ",
}


class SearchService:
    def search(self, query: str) -> list[SearchResult]:
        query = query.strip()
        if not query:
            return []

        results: list[SearchResult] = []
        seen_symbols: set[str] = set()

        # 1. Try exact ticker lookup
        self._try_exact_match(query.upper(), results, seen_symbols)

        # 2. Try with common exchange suffixes for non-US markets
        for exchange, suffix in EXCHANGE_SUFFIXES.items():
            if suffix:
                symbol_with_suffix = f"{query.upper()}{suffix}"
                self._try_exact_match(
                    symbol_with_suffix, results, seen_symbols, exchange
                )

        # 3. Use yfinance search for broader matching (by company name)
        try:
            search_obj = Search(query, max_results=MAX_SEARCH_RESULTS)
            quotes = search_obj.quotes or []
            for quote in quotes:
                symbol = quote.get("symbol", "")
                if symbol and symbol not in seen_symbols:
                    seen_symbols.add(symbol)
                    results.append(
                        SearchResult(
                            symbol=symbol,
                            name=quote.get("longname", quote.get("shortname", symbol)),
                            exchange=quote.get("exchDisp", quote.get("exchange", "Unknown")),
                            country=quote.get("country", _guess_country(symbol)),
                            market_cap=quote.get("marketCap"),
                        )
                    )
        except Exception:
            pass

        # Sort by market cap (largest first), nulls last
        results.sort(
            key=lambda r: r.market_cap if r.market_cap is not None else -1,
            reverse=True,
        )

        return results[:MAX_SEARCH_RESULTS]

    def _try_exact_match(
        self,
        symbol: str,
        results: list[SearchResult],
        seen: set[str],
        exchange_hint: str | None = None,
    ) -> None:
        if symbol in seen:
            return
        try:
            ticker = yf.Ticker(symbol)
            info = ticker.info
            if info and info.get("regularMarketPrice") is not None:
                seen.add(symbol)
                results.append(
                    SearchResult(
                        symbol=symbol,
                        name=info.get("longName", info.get("shortName", symbol)),
                        exchange=exchange_hint or info.get("exchange", "Unknown"),
                        country=info.get("country", _guess_country(symbol)),
                        market_cap=info.get("marketCap"),
                    )
                )
        except Exception:
            pass


def _guess_country(symbol: str) -> str:
    """Guess country from exchange suffix."""
    suffix_country = {
        ".L": "United Kingdom",
        ".PA": "France",
        ".AS": "Netherlands",
        ".DE": "Germany",
        ".CO": "Denmark",
        ".ST": "Sweden",
        ".HE": "Finland",
        ".T": "Japan",
        ".HK": "Hong Kong",
        ".SS": "China",
        ".SZ": "China",
    }
    for suffix, country in suffix_country.items():
        if symbol.endswith(suffix):
            return country
    return "United States"

File: frontend/src/lib/services/searchService.ts
/**
 * Search service for finding stocks and ETFs.
 * Uses yahoo-finance2 for search functionality.
 */

import YahooFinance from "yahoo-finance2";
const yahooFinance = new YahooFinance();
import { EXCHANGE_SUFFIXES, MAX_SEARCH_RESULTS } from "../config";
import type { SearchResult } from "@/types";

// Guess country from exchange suffix
function guessCountry(symbol: string): string {
  const suffixCountry: Record<string, string> = {
    ".L": "United Kingdom",
    ".PA": "France",
    ".AS": "Netherlands",
    ".DE": "Germany",
    ".CO": "Denmark",
    ".ST": "Sweden",
    ".HE": "Finland",
    ".T": "Japan",
    ".HK": "Hong Kong",
    ".SS": "China",
    ".SZ": "China",
  };

  for (const [suffix, country] of Object.entries(suffixCountry)) {
    if (symbol.endsWith(suffix)) {
      return country;
    }
  }
  return "United States";
}

// Try exact match for a symbol
async function tryExactMatch(
  symbol: string,
  results: SearchResult[],
  seen: Set<string>,
  exchangeHint?: string
): Promise<void> {
  if (seen.has(symbol)) return;

  try {
    const quote: any = await yahooFinance.quote(symbol);
    if (quote && quote.regularMarketPrice !== undefined) {
      seen.add(symbol);
      results.push({
        symbol,
        name: quote.longName || quote.shortName || symbol,
        exchange: exchangeHint || quote.exchange || "Unknown",
        country: guessCountry(symbol),
        market_cap: quote.marketCap || null,
        is_etf: quote.quoteType === "ETF",
      });
    }
  } catch {
    // Symbol not found, ignore
  }
}

export async function searchStocks(query: string): Promise<SearchResult[]> {
  const trimmedQuery = query.trim();
  if (!trimmedQuery) {
    return [];
  }

  const results: SearchResult[] = [];
  const seen = new Set<string>();

  // 1. Try exact ticker lookup
  await tryExactMatch(trimmedQuery.toUpperCase(), results, seen);

  // 2. Try with common exchange suffixes for non-US markets
  for (const [exchange, suffix] of Object.entries(EXCHANGE_SUFFIXES)) {
    if (suffix) {
      const symbolWithSuffix = `${trimmedQuery.toUpperCase()}${suffix}`;
      await tryExactMatch(symbolWithSuffix, results, seen, exchange);
    }
  }

  // 3. Use yahoo-finance2 search for broader matching
  try {
    const searchResults: any = await yahooFinance.search(trimmedQuery, {
      newsCount: 0,
      quotesCount: MAX_SEARCH_RESULTS,
    });

    if (searchResults?.quotes) {
        for (const quote of searchResults.quotes as any[]) {
        // Filter to only equity and ETF types
        if (quote.quoteType !== "EQUITY" && quote.quoteType !== "ETF") {
          continue;
        }

        const symbol = quote.symbol;
        if (symbol && !seen.has(symbol)) {
          seen.add(symbol);

          // Get more details with a quote call
          try {
                    const detailedQuote: any = await yahooFinance.quote(symbol);
            results.push({
              symbol,
              name: detailedQuote?.longName || detailedQuote?.shortName || quote.shortname || symbol,
              exchange: detailedQuote?.exchange || quote.exchange || "Unknown",
              country: guessCountry(symbol),
              market_cap: detailedQuote?.marketCap || null,
              is_etf: quote.quoteType === "ETF",
            });
          } catch {
            // Use basic info from search if quote fails
            results.push({
              symbol,
              name: quote.shortname || symbol,
              exchange: quote.exchange || "Unknown",
              country: guessCountry(symbol),
              market_cap: null,
              is_etf: quote.quoteType === "ETF",
            });
          }
        }

        // Limit results to avoid too many API calls
        if (results.length >= MAX_SEARCH_RESULTS) {
          break;
        }
      }
    }
  } catch {
    // Search failed, return what we have
  }

  // Sort by market cap (largest first), nulls last
  results.sort((a, b) => {
    if (a.market_cap === null && b.market_cap === null) return 0;
    if (a.market_cap === null) return 1;
    if (b.market_cap === null) return -1;
    return b.market_cap - a.market_cap;
  });

  return results.slice(0, MAX_SEARCH_RESULTS);
}
